### ISSUE TITLE
@JsonTypeInfo on an interface does not work

### ISSUE THREAD

**Issue by richturner** [victools/jsonschema-generator#185]
`JsonSubTypesResolver` navigates up the type hierarchy to find a super type with `JsonTypeInfo` annotation but it only looks at superclass and not interfaces:

https://github.com/victools/jsonschema-generator/blob/1c7b7a8830eed60aafe850115776dfb419b8bede/jsonschema-module-jackson/src/main/java/com/github/victools/jsonschema/module/jackson/JsonSubTypesResolver.java#L112

It's quite common to use `JsonTypeInfo` on an interface so interfaces should be searched, guess there's a risk of selecting the wrong interface and ideally the super type for which schema is being generated should be exposed to the resolver so the resolver doesn't have to go searching through the type hierarchy.

As a side note I wonder if the jackson object mapper could be utilised for such type info extraction.

**Comment by CarstenWickner** [comment]
Hi @richturner,

Thanks for pointing that out.
I'll have a look at whether that's easy enough to fix.

As to "using the Jackson `ObjectMapper`", the challenge with that is typically the different scope.
- The `ObjectMapper` looks at an instance.
- The schema generator looks only at a static type.

In the beginning, I spent considerable hours/days trying to make the type resolution work with what Jackson does internally, until I resorted to using `classmate` (from the main maintainer of Jackson), as that is more suited for static type resolution (although it also has its limits).

**Comment by CarstenWickner** [comment]
Hi @richturner,

I agree with your assessment that this backward search should ideally be avoided, but I'm shying away from that extent of refactoring right now – due to a lack of time on my part.
I've merged a simple fix together with a couple of additional unit tests. That'd work for simple scenarios where there are no conflicting type declarations on multiple interfaces.

Not releasing it yet, as I'd actually prefer to do the more extensive refactoring in order to solve this cleanly.
I'm unsure when I'll have the time for it though.

**Comment by CarstenWickner** [comment]
Thinking about it some more, when a particular subtype is declared as variable, it is still being wrapped/augmented according to the `@JsonTypeInfo` annotation on one of its parents. I.e., the refactoring would only help in some cases but not in others.

I've briefly tested it and Jackson appears to be taking the first interface with a `@JsonTypeInfo` annotation on it.
This test:
```java
public class SubtypeResolutionConflictText {

    @Test
    public void testConflictResolution() throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        System.out.println("AB: " + mapper.writeValueAsString(new AB()));
        System.out.println("BA: " + mapper.writeValueAsString(new BA()));
    }

    static class AB implements A, B {
    }
    static class BA implements B, A {
    }

    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.WRAPPER_OBJECT)
    @JsonSubTypes({
        @JsonSubTypes.Type(value = AB.class, name = "AB"),
        @JsonSubTypes.Type(value = BA.class, name = "BA")
    })
    interface A {
    }

    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY)
    @JsonSubTypes({
        @JsonSubTypes.Type(value = AB.class, name = "AB"),
        @JsonSubTypes.Type(value = BA.class, name = "BA")
    })
    interface B {
    }
}
```
produces the following output:
```
AB: {"AB":{}}
BA: {"@type":"BA"}
```

The current/simple implementation seems to be alright then. I still hesitate to release it quite yet.

**Comment by richturner** [comment]
Thanks for taking the time to look into this issue; I guess the scenario of having multiple `@JsonTypeInfo` is a strange one to handle and going with the first encountered makes sense beyond 'failing hard and fast' doesn't seem to be the Jackson way.

**Comment by CarstenWickner** [comment]
This has been released in [v4.19.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.19.0) now.
